2_process.smk

```diff
@@ -0,0 +1,155 @@
+        num_sequences = np.load(sequences_file).shape[0] 
+        sequence_counts.append(num_sequences)
+    df_counts = pd.DataFrame(data={
+        'sequences_file': lake_sequence_files_input,
+        'num_sequences': sequence_counts
+    })
+    df_counts.to_csv(summary_file, index=False)
+
+
+def dynamic_filenames(wildcards):
+    """
+    Return the files that contain dynamic data that are needed to construct
+    sequences for a given lake.
+
+    This function also triggers four checkpoints:
+    1. fetch_mntoha_metadata to get lake_metadata.csv
+    2. unzip_archive for this lake's drivers
+    3. unzip_archive for this lake's clarity
+    3. unzip_archive for this lake's ice flags
+
+    :param wildcards: Snakemake wildcards (site_id from mntoha_lake_sequences).
+    :returns: List of 3 dynamic filenames: drivers, clarity, and ice flags
+
+    """
+    # make this function depend on fetch_mntoha_metadata
+    # needed because lake_metadata.csv is used to determine dynamic files
+    lake_metadata_file = checkpoints.fetch_mntoha_metadata.get(**wildcards).output[0]
+    lake_metadata = pd.read_csv(lake_metadata_file)
+    lake = lake_metadata.loc[lake_metadata['site_id']==wildcards.site_id].iloc[0]
+    # also make this function depend on unzip_archive
+    # needed to link unzipped files with unzip_archive rule
+    file_category = 'dynamic_mntoha'
+    drivers_directory = f'inputs_{lake.group_id}'
+    unzip_archive_drivers = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=drivers_directory
+    ).output[0]
+    clarity_directory = f'clarity_{lake.group_id}'
+    unzip_archive_clarity = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=clarity_directory
+    ).output[0]
+    ice_flags_directory = f'ice_flags_{lake.group_id}'
+    unzip_archive_ice_flags = checkpoints.unzip_archive.get(
+        file_category=file_category,
+        directory_name=ice_flags_directory
+    ).output[0]
+
+    # dynamic filenames
+    dynamic_dir = '1_fetch/out'
+    drivers_file = f'{dynamic_dir}/dynamic_mntoha/inputs_{lake.group_id}/{lake.meteo_filename}'
+    clarity_file = f'{dynamic_dir}/dynamic_mntoha/clarity_{lake.group_id}/gam_{lake.site_id}_clarity.csv'
+    ice_flags_file = f'{dynamic_dir}/dynamic_mntoha/ice_flags_{lake.group_id}/pb0_{lake.site_id}_ice_flags.csv'
+    return [drivers_file, clarity_file, ice_flags_file]
+
+
+# Summarize training sequences
+rule process_mntoha:
+    input:
+        mntoha_lake_sequence_files
+    output:
+        "2_process/out/mntoha_sequences/mntoha_sequences_summary.csv"
+    run:
+        save_sequences_summary(input, output[0])
+
+
+# Create .npy of input/output sequences for one lake to use for training and testing
+rule mntoha_lake_sequences:
+    input:
+        "2_process/tmp/mntoha/lake_metadata_augmented.csv",
+        "2_process/tmp/mntoha/temperature_observations_interpolated.csv",
+        dynamic_filenames
```

Comment:
For the sake of readability and avoiding hard-coding within functions I find myself wishing there were some way to specify parameters for these functions that make use of checkpoints here in the `smk` file -- e.g., that file_category = 'dynamic_mntoha'.

Reply:
From the docs [here](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#input-functions-and-unpack), it looks like this can be achieved. Rather than passing the function `dynamic_mntoha`, we could try calling the function and using Python's star operator to unpack the list of file names, like this:
```
input:
    "2_process/tmp/mntoha/lake_metadata_augmented.csv",
    "2_process/tmp/mntoha/temperature_observations_interpolated.csv",
    *dynamic_filenames(file_category='dynamic_mntoha')
```

Reply:
Indeed, wildcards weren't passed. Instead, using a `lambda` function allowed a keyword to be specified. 
```
    input:
        "2_process/tmp/mntoha/lake_metadata_augmented.csv",
        "2_process/tmp/mntoha/temperature_observations_interpolated.csv",
        lambda wildcards: dynamic_filenames(wildcards.site_id, file_category='dynamic_mntoha')
```
Also, then, the function `dynamic_filenames` doesn't have to bother with the `wildcards` variable. The arguments are now simply the arguments that `dynamic_filenames` needs.

Reply:
Ooh nice pipelines upgrade, Andy!